lib/FreeRTOS/mqtt-agent-interface/freertos_command_pool.c

```diff
@@ -0,0 +1,142 @@
+ */
+
+/* Standard includes. */
+#include <string.h>
+#include <stdio.h>
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "semphr.h"
+
+/* Header include. */
+#include "freertos_command_pool.h"
+#include "freertos_agent_message.h"
+
+/*-----------------------------------------------------------*/
+
+#define QUEUE_NOT_INITIALIZED    ( 0U )
+#define QUEUE_INITIALIZED        ( 1U )
+
+/**
+ * @brief The pool of command structures used to hold information on commands (such
+ * as PUBLISH or SUBSCRIBE) between the command being created by an API call and
+ * completion of the command by the execution of the command's callback.
+ */
+static Command_t commandStructurePool[ MQTT_COMMAND_CONTEXTS_POOL_SIZE ];
+
+/**
+ * @brief A queue used to guard the pool of Command_t structures. Structures
+ * may be obtained by receiving a pointer from the queue, and returned by
+ * sending the pointer back into it.
+ */
+static AgentMessageContext_t commandStructQueue;
+
+/**
+ * @brief Initialization status of the queue.
+ */
+static volatile uint8_t initStatus = QUEUE_NOT_INITIALIZED;
+
+/*-----------------------------------------------------------*/
+
+void Agent_InitializePool( void )
+{
+    size_t i;
+    Command_t * pCommand;
+    static uint8_t staticQueueStorageArea[ MQTT_COMMAND_CONTEXTS_POOL_SIZE * sizeof( Command_t * ) ];
+    static StaticQueue_t staticQueueStructure;
+    bool commandAdded = false;
+
+    if( initStatus == QUEUE_NOT_INITIALIZED )
+    {
+        memset( ( void * ) commandStructurePool, 0x00, sizeof( commandStructurePool ) );
+        commandStructQueue.queue = xQueueCreateStatic( MQTT_COMMAND_CONTEXTS_POOL_SIZE,
+                                                       sizeof( Command_t * ),
+                                                       staticQueueStorageArea,
+                                                       &staticQueueStructure );
+        configASSERT( commandStructQueue.queue );
+
+        /* Populate the queue. */
+        for( i = 0; i < MQTT_COMMAND_CONTEXTS_POOL_SIZE; i++ )
+        {
+            /* Store the address as a variable. */
+            pCommand = &commandStructurePool[ i ];
+            /* Send the pointer to the queue. */
+            commandAdded = Agent_MessageSend( &commandStructQueue, &pCommand, 0U );
+            configASSERT( commandAdded );
+        }
+
+        initStatus = QUEUE_INITIALIZED;
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+Command_t * Agent_GetCommand( uint32_t blockTimeMs )
+{
+    Command_t * structToUse = NULL;
+    size_t i;
+    bool structRetrieved = false;
+
+    /* Check queue has been created. */
+    if( initStatus == QUEUE_INITIALIZED )
+    {
+        /* Retrieve a struct from the queue. */
+        structRetrieved = Agent_MessageReceive( &commandStructQueue, &( structToUse ), blockTimeMs );
+
+        if( !structRetrieved )
+        {
+            LogError( ( "No command structure available." ) );
+        }
+    }
```

Comment:
It will be better to have a error log here if the Queue is not initialized, so that the application developer can easily find the root cause if a call to Agent_InitializePool is missed.

Reply:
I'll just add an assert instead, as the queue really needs to be initialized

Reply:
Sounds good.